qiskit/transpiler/passes/basis/decompose.py

```diff
@@ -85,19 +88,25 @@ def run(self, dag: DAGCircuit) -> DAGCircuit:
             output dag where ``gate`` was expanded.
         """
         # Walk through the DAG and expand each non-basis node
-        for node in dag.op_nodes():
-            if self._should_decompose(node):
-                if node.op.definition is None:
-                    continue
-                # TODO: allow choosing among multiple decomposition rules
-                rule = node.op.definition.data
-                if len(rule) == 1 and len(node.qargs) == len(rule[0][1]) == 1:
-                    if node.op.definition.global_phase:
-                        dag.global_phase += node.op.definition.global_phase
-                    dag.substitute_node(node, rule[0][0], inplace=True)
-                else:
-                    decomposition = circuit_to_dag(node.op.definition)
-                    dag.substitute_node_with_dag(node, decomposition)
+        for _ in range(self.reps):
+            gates = []
+            for node in dag.op_nodes():
+                if self._should_decompose(node):
+                    if node.op.definition is None:
+                        continue
+                    # TODO: allow choosing among multiple decomposition rules
+                    rule = node.op.definition.data
+                    if len(rule) == 1 and len(node.qargs) == len(rule[0][1]) == 1:
+                        if node.op.definition.global_phase:
+                            dag.global_phase += node.op.definition.global_phase
+                        dag.substitute_node(node, rule[0][0], inplace=True)
+                        gates.append(node.op.name)
+                    else:
+                        decomposition = circuit_to_dag(node.op.definition)
+                        dag.substitute_node_with_dag(node, decomposition)
+                        for nod in dag.op_nodes():
+                            gates.append(nod.name)
+            self.gates_to_decompose = gates
```

Comment:
The pass really shouldn't be modifying `gates_to_decompose`: that's a stateful parameter, and this makes the pass unusable after one run.

Reply:
I don't understand what is unusable mean, may be add a test for it. my idea is that, it could repeatly decompose the gates_to_decompose, so it shouldn't effect anything in ``gates_to_decompose`` after a run, I check it with 
`dag = circuit_to_dag(qc.decompose('gate*',reps=1))
gates_to_decompose = []
for node in dag.op_nodes():
    gates_to_decompose.append(node.name)`